pkg/dynatrace-client/client.go

```diff
@@ -269,3 +294,54 @@ func readHostMap(r io.Reader) (map[string]string, error) {
 
 	return result, nil
 }
+
+// GetCommunicationHosts returns the list of communication hosts used on communication endpoints
+// for the environment.
+func readCommunicationHosts(r io.Reader) ([]CommunicationHost, error) {
+	type jsonResponse struct {
+		CommunicationEndpoints []string
+
+		Error *serverError
+	}
+
+	var resp jsonResponse
+	switch err := json.NewDecoder(r).Decode(&resp); {
+	case err != nil:
+		return nil, err
+	case resp.Error != nil:
+		return nil, resp.Error
+	}
+
+	out := make([]CommunicationHost, len(resp.CommunicationEndpoints))
+
+	for i, s := range resp.CommunicationEndpoints {
+		u, err := url.ParseRequestURI(s)
+		if err != nil {
+			return nil, err
+		}
+
+		rp := u.Port() // Empty if not included in the URI
+		if rp == "" {
+			switch u.Scheme {
+			case "http":
+				rp = "80"
+			case "https":
+				rp = "443"
+			default:
+				return nil, fmt.Errorf("Unknown scheme: %s", u.Scheme)
+			}
+		}
+
+		p, err := strconv.Atoi(rp)
+		if err != nil {
+			return nil, err
+		}
+
+		out[i] = CommunicationHost{
+			Host: u.Hostname(),
+			Port: p,
+		}
+	}
```

Comment:
Thought about it. The cause would be malformed AG URLs coming from the environment. Well, perhaps if the URL is a file, but I doubt that is an scenario.

We could send a logger as a parameter to the function, and log the failures, while also skipping the entry.

Reply:
Made the changes to use a static logger rather than passing it as a parameter. We'd be missing the context values used in the Reconcile() logger, but they aren't relevant for the client messages ATM.